tree/ntuple/v7/src/RNTupleProcessor.cxx

```diff
@@ -352,135 +349,122 @@ ROOT::Experimental::RNTupleJoinProcessor::RNTupleJoinProcessor(const RNTupleOpen
+         return fieldName.substr(0, n.fNTupleName.size()) == n.fNTupleName;
+      });
+      if (isAuxParent != auxNTuples.end())
+         continue;
+
       // If the model provided by the user has a default entry, use the value pointers from the default entry of the
       // model that was passed to this constructor. This way, the pointers returned by RNTupleModel::MakeField can be
       // used in the processor loop to access the corresponding field values.
-      if (!models.empty() && !models[0]->IsBare()) {
-         auto valuePtr = models[0]->GetDefaultEntry().GetPtr<void>(fieldName);
+      if (!fModel->IsBare()) {
+         auto valuePtr = fModel->GetDefaultEntry().GetPtr<void>(fieldName);
          fEntry->BindValue(fieldName, valuePtr);
       }
 
-      const auto &[fieldContext, _] =
-         fFieldContexts.try_emplace(fieldName, field.Clone(fieldName), fEntry->GetToken(fieldName));
-      ConnectField(fieldContext->second, *fPageSource, *fEntry);
+      fFieldContexts.try_emplace(fieldName, field.Clone(fieldName), fEntry->GetToken(fieldName));
    }
 
    for (unsigned i = 0; i < auxNTuples.size(); ++i) {
-      if (models.empty()) {
-         AddAuxiliary(auxNTuples[i], joinFields);
-      } else {
-         // The size of `models` is checked in `CreateJoin`; at this point we can safely assume that `models.size() ==
-         // auxNTuples.size() + 1`.
-         AddAuxiliary(auxNTuples[i], joinFields, std::move(models[i + 1]));
-      }
+      AddAuxiliary(auxNTuples[i], joinFields, i + 1 /* ntupleIdx */);
    }
 }
 
-void ROOT::Experimental::RNTupleJoinProcessor::AddAuxiliary(const RNTupleOpenSpec &auxNTuple,
-                                                            const std::vector<std::string> &joinFields,
-                                                            std::unique_ptr<RNTupleModel> model)
+void ROOT::Experimental::RNTupleJoinProcessor::SetModel(std::unique_ptr<RNTupleModel> primaryModel,
+                                                        std::vector<std::unique_ptr<RNTupleModel>> auxModels)
 {
-   assert(fNEntriesProcessed == 0 && "cannot add auxiliary ntuples after processing has started");
-
-   fNTuples.emplace_back(auxNTuple);
-
-   auto pageSource = Internal::RPageSource::Create(auxNTuple.fNTupleName, auxNTuple.fStorage);
-   pageSource->Attach();
-
-   if (pageSource->GetNEntries() == 0) {
-      throw RException(R__FAIL("provided RNTuple is empty"));
-   }
-
-   if (!model)
-      model = pageSource->GetSharedDescriptorGuard()->CreateModel();
+   fModel = std::move(primaryModel);
+   fModel->Unfreeze();
 
-   model->Freeze();
-   auto entry = model->CreateBareEntry();
+   // Create an anonymous record field for each auxiliary ntuple, containing their top-level fields. These original
+   // top-level fields are registered as subfields in the join model, such that they can be accessed as
+   // `auxNTupleName.fieldName`.
+   // TODO(fdegeus) Support projected fields
```

Comment:
I'm not sure what this means. IIRC projected fields are generally not "supported" for reading, ie they are duplicated in the entry

Reply:
You're right, honestly I don't remember my train of thought when adding this comment and it works all fine. I'll delete it!